buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.gradlebuild.buildquality.quick
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.tasks.TaskAction
+import org.gradle.cache.internal.GeneratedGradleJarCache
+import org.gradle.kotlin.dsl.*
+import org.gradle.kotlin.dsl.experiments.plugins.GradleKotlinDslKtlintConventionPlugin
+import org.gradle.kotlin.dsl.support.serviceOf
+import java.util.concurrent.Callable
+
+
+class QuickCheckPlugin : Plugin<Project> {
+    override fun apply(project: Project) {
+        project.tasks.register("quickCheck", QuickCheckTask::class.java)
+    }
+}
+
+
+abstract class QuickCheckTask() : DefaultTask() {
+    @TaskAction
+    fun execute() {
+        if (System.getenv("IGNORE_QUICK_CHECK") != null) {
+            return
+        }
+        val changedFiles = getChangedFiles()
+
+        project.configurations.create("quickCheck")
+        val checks = Check.values().filter { it.filter(changedFiles).isNotEmpty() }
+        checks.forEach { it.addDependencies(project) }
+        checks.forEach { it.runCheck(project, it.filter(changedFiles)) }
+    }
+
+    //    A       buildSrc/subprojects/buildquality/src/main/kotlin/org/gradle/gradlebuild/buildquality/quick/QuickCheck.kt
+    //    M       gradle.properties
+    //    D       gradle/Check.java
+    private
+    fun getChangedFiles(): List<String> =
+        project.execAndGetStdout("git", "diff", "--cached", "--name-status", "HEAD")
```

Comment:
Would using `jgit` make sense here instead of relying on CLI?

Reply:
`jgit` would be several hundred ms faster. But CLI is more readable and maintable. So let's keep CLI unless it becomes a problem.

Reply:
I suppose you should use `--porcelain`, so the output can be parsed reliably.

Reply:
Maybe I misunderstand you, but `git status --porcelain` has same output as `git diff --cached --name-status HEAD`:

```
M build.gradle.kts
A some-new-file
```

Why `--porcelain` are more reliable?

Reply:
This is the documentation: https://git-scm.com/docs/git-status#Documentation/git-status.txt---p

The "porcelain" commands (like `checkout`, `pull`, `commit`, etc.) have no guarantee for stable output between versions. The `--porcelain` command line flag adds this guarantee, so the output is easy to parse for scripts.

Reply:
Thanks @wolfs